sbws/lib/destination.py

```diff
@@ -91,80 +99,166 @@ def connect_to_destination_over_circuit(dest, circ_id, session, cont, max_dl):
-        # (which might be the relays fault, not the destination being
-        # unreachable)
-        # - the destination does not support HTTP Range requests.
-        self.failed = False
-        self.consecutive_failures = 0
 
-    @property
-    def is_functional(self):
+        # Attributes to decide whether a destination is functional or not.
+        self._max_num_failures = max_num_failures
+        self._num_attempts_stored = num_attempts_stored
+        # Default delta time to try a destination that was not functional.
+        self._default_delta_seconds_retry = delta_seconds_retry
+        self._delta_seconds_retry = delta_seconds_retry
+        # Using a deque (FIFO) to do not grow forever and
+        # to do not have to remove old attempts.
+        # Store tuples of timestamp and whether the destination succed or not
+        # (succed, 1, failed, 0).
+        # Initialize it as if it never failed.
+        self._attempts = collections.deque([(datetime.datetime.utcnow(), 1), ],
+                                           maxlen=self._num_attempts_stored)
+        self._factor = factor_increment_retry
+
+    def _last_attempts(self, n=None):
+        """Return the last ``n`` attempts the destination was used."""
+        # deque does not accept slices,
+        # a new deque is returned with the last n items
+        # (or less if there were less).
+        return collections.deque(self._attempts,
+                                 maxlen=(n or self._max_num_failures))
+
+    def _are_last_attempts_failures(self, n=None):
+        """
+        Return True if the last`` n`` times the destination was used
+        and failed.
         """
-        Returns True if there has not been a number consecutive measurements.
-        Otherwise warn about it and return False.
+        # Count the number that there was a failure when used
+        n = n if n else self._max_num_failures
+        return ([i[1] for i in self._last_attempts(n)].count(0)
+                >= self._max_num_failures)
 
+    def _increment_time_to_retry(self, factor=None):
         """
-        if self.consecutive_failures > MAXIMUM_NUMBER_DESTINATION_FAILURES:
-            log.warning("Destination %s is not functional. Please check that "
-                        "it is correct.", self._url)
+        Increment the time a destination will be tried again by a ``factor``.
+        """
+        self._delta_seconds_retry *= factor or self._factor
+        log.info("Incremented the time to try destination %s to %s hours.",
+                 self.url, self._delta_seconds_retry / 60 / 60)
+
+    def _is_last_try_old_enough(self, n=None):
+        """
+        Return True if the last time it was used it was ``n`` seconds ago.
+        """
+        # Timestamp of the last attempt.
+        last_time = self._attempts[-1][0]
+        # If the last attempt is older than _delta_seconds_retry,
+        if (datetime.datetime.utcnow()
```

Comment:
This can be shortened to `return (datetime.datetime.utcnow() - datetime.timedelta(seconds=self._delta_seconds_retry)) > last_time`